### DISCUSSION TITLE
Making a mod to change the HitStatus of a circle hit

### DISCUSSION THREAD

**Discussion by leryss** [ppy/osu#13222] [Q&A] [answered]
Im trying to implement a mod that treats HitResult.Meh, HitResult.Good, HitResult.Ok as HitResult.Miss, basically Perfect mod but without instantly failing the map.

The only solution I could find was to implement IApplicableToDrawableHitObjects and register a callback for the OnNewResult event for each drawable hit object, in this callback I simply check the judgement type and replace the type if necessary

```csharp
public void ApplyToDrawableHitObjects(IEnumerable<DrawableHitObject> drawables)
{
    if (!PerfectHitsOnly.Value)
        return;

    foreach (var drawable in drawables)
    {
        drawable.OnNewResult += onNewJudgement;
    }
}

private void onNewJudgement(DrawableHitObject hitObject, JudgementResult judgement)
{
    if (judgement.Type > HitResult.Miss && judgement.Type < HitResult.Great)
    {
        judgement.Type = HitResult.Miss;
    }
}
```

For this solution to work I had however to change the location where the event in DrawableHitObject.cs otherwise the wrong state would be updated when I mutate the JudgementResult

```csharp
protected void ApplyResult(Action<JudgementResult> application)
{
   ...

    OnNewResult?.Invoke(this, Result);

    if (Result.HasResult)
        updateState(Result.IsHit ? ArmedState.Hit : ArmedState.Miss);

    // Here was the original OnNewResult?.Invoke
}
```

@peppy told me this is bad so I would be better off starting a discussion on this topic to find another solution. Any thoughts ?

**Comment by frenzibyte** [comment, marked as answer]
That is without doubt a bad thing to do, events must be the last thing to invoke after the internals of the object have been updated in line with the given new judgement result, that's what's to be expected in here.

The idea of your mod sounds like it's looking for a `HitWindow` that ignores all other hit results but `HitResult.Great`, but it'll have to depend on https://github.com/ppy/osu/issues/12918 being resolved first and foremost.

**Reply by peppy** [reply]
> events must be the last thing to invoke after the internals of the object have been updated in line with the given new judgement result, that's what's to be expected in here.

This is a bit incorrect (there's no specification/rule that says this). The issue I brought up in the original PR is that this was *repurposing an event* that was intended to be exposing the final state. There's nothing stopping us from adding a hook to allow judgement mutation from mods earlier in the proposed method.

**Reply by frenzibyte** [reply]
Perhaps I got over the top there, I meant that for `OnXYZ` events specifically, and stated it as a known and expected thing for such events. There are already existing events that mutate the object (e.g. `ApplyCustomUpdateState`), so I wouldn't be against having something like that either.

**Reply by leryss** [reply]
ApplyCustomUpdateState looks like what I would need, the problem is that it doesnt allow changing the ArmedState, this results in hit sounds still playing even if judgement is Miss. By the looks of it quite some work is needed to allow changing of ArmedState. So what peppy suggested I think would work the best, an event to override the judgement right before the state update.

**Reply by smoogipoo** [reply]
Using `HitWindows` is the right idea, I think. May run into https://github.com/ppy/osu/issues/12918 though.
